### PR TITLE
Use the initial peer success count for the fanout limit

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra is using the total number of seed peers for the fanout limit.

The limit will be too high if any of the initial peer connections fail.
(Typically, lots of those connections fail.)

## Solution

- Use the initial peer success count for the fanout limit
- Make sure that the crawler starts after the first successful peer (or after all the initial handshake attempts have failed)

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Integration Tests

## Review

Whoever reviews #2181 should also review this task. (@dconnolly or @jvff)

## Related Issues

Part of #2163

Based on PR #2181 
Split off PR #2160

## Follow Up Work

Rest of #2160

### CODE REVIEW

**Comment by teor2345** [comment]
I've put this PR in draft until #2181 merges to main.

File: zebra-network/src/peer_set/initialize.rs [hunk 1]
```diff
@@ -166,9 +171,13 @@ where
...
 
-    info!("Sending initial request for peers");
+    info!("waiting for a successful initial peer connection");
+    // it doesn't matter if the sender has been dropped
+    let _ = initial_success_count_rx.changed().await;
+    info!(initial_successes = ?*initial_success_count_rx.borrow(),
+          "asking initial peers for new peers");
     let _ = candidates
-        .update_initial(initial_peer_count_rx.await.expect("value sent before drop"))
+        .update_initial(*initial_success_count_rx.borrow())
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:180]: Use the initial peer success count for the fanout limit, Part 2
- Wait for the first success count update
- Start the crawler using that success count

File: zebra-network/src/peer_set/initialize.rs [hunk 2]
```diff
@@ -284,16 +296,19 @@ where
                     // the peer set is handled by an independent task, so this send
                     // shouldn't hang
                     peerset_tx.send(Ok(peer_set_change)).await?;
+                    // if the receiver has been dropped, we still want to process
+                    // the handshakes
+                    let _ = success_count_tx.send(success_count);
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:301]: Use the initial peer success count for the fanout limit, Part 1a
- Send the success count whenever we have a success
- This also starts the crawler

File: zebra-network/src/peer_set/initialize.rs [hunk 3]
```diff
@@ -329,8 +344,10 @@ where
         warn!(
             ?initial_peers_len,
             ?peerset_initial_target_size,
-            "no successful initial peer connections"
+            "no successful initial peer connections, starting crawler anyway"
         );
+        // this redundant update will start the crawler
+        let _ = success_count_tx.send(success_count);
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:350]: Use the initial peer success count for the fanout limit, Part 1b

- Start the crawler anyway if all connections have been attempted

Comment (jvff) [reply zebra-network/src/peer_set/initialize.rs:350]: Is this necessary? At the end of the method `success_count_tx` will get dropped, resuming the `changed()` future anyway, right?

**Review by jvff — APPROVED** [review]

File: zebra-network/src/peer_set/initialize.rs (hunk 3 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:350]: I think you're right, I wasn't really paying attention to the error.

Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:350]: This code is removed in the latest #2160.

**Comment by teor2345** [comment]
This bug is fixed in #2160, but since #2160 integrates the initial handshakes into the crawler, the code is quite different.

I'll open a new PR for it.

**Pull request closed without merging**